criu/files-reg.c

```diff
@@ -1005,11 +1006,21 @@ static int create_link_remap(char *path, int len, int lfd, u32 *idp, struct ns_i
 	mntns_root = mntns_get_root_fd(nsid);
 
 again:
-	ret = linkat_hard(lfd, "", mntns_root, link_name, st->st_uid, st->st_gid, AT_EMPTY_PATH);
+	ret = linkat_hard(lfd, "", mntns_root, link_name, ost->st_uid, ost->st_gid, AT_EMPTY_PATH);
 	if (ret < 0 && errno == ENOENT) {
 		/* Use grand parent, if parent directory does not exist. */
 		if (trim_last_parent(link_name) < 0) {
 			pr_err("trim failed: @%s@\n", link_name);
+			if (fb && parms->fs_type == OVERLAYFS_SUPER_MAGIC) {
+				/* In overlayFS, linkat() fails with ENOENT if the removed file is originated from lower layer.
+				 * The cause of failure is that linkat() sees the file has st_nlink=0,
+				 * which is different than st_nlink=1 we got from earlier fstat() on lfd.
+				 * By setting *fb=true, we will fall back to dump_ghost_remap() as it is what should have been
+				 * done to removed files with st_nlink=0.
+				 */
+				pr_info("Unable to link-remap %s on overlayFS, fall back to dump_ghost_remap\n", path);
```

Comment:
nit: We are following the kernel coding style and it limits the line size to 120 symbols. It can be reworked like this:

diff --git a/criu/files-reg.c b/criu/files-reg.c
index b58aed9aa..099b6af93 100644
--- a/criu/files-reg.c
+++ b/criu/files-reg.c
@@ -959,7 +959,25 @@ void free_link_remaps(void)
 }
 static int linkat_hard(int odir, char *opath, int ndir, char *npath, uid_t uid, gid_t gid, int flags);
 
-static int create_link_remap(char *path, int len, int lfd, u32 *idp, struct ns_id *nsid, const struct fd_parms *parms, bool *fb)
+static void check_overlayfs_fallback(const struct fd_parms *params, bool *fallback) {
+       if (!fallback || parms->fs_type != OVERLAYFS_SUPER_MAGIC)
+               return;
+
+       /*
+        * In overlayFS, linkat() fails with ENOENT if the removed file is
+        * originated from lower layer.  The cause of failure is that linkat()
+        * sees the file has st_nlink=0, which is different than st_nlink=1 we
+        * got from earlier fstat() on lfd.  By setting *fb=true, we will fall
+        * back to dump_ghost_remap() as it is what should have been done to
+        * removed files with st_nlink=0.
+        */
+       pr_info("Unable to link-remap %s on overlayFS, fall back to dump_ghost_remap\n", path);
+       *fallback = true;
+}
+
+static int create_link_remap(char *path, int len, int lfd, u32 *idp,
+                            struct ns_id *nsid, const struct fd_parms *parms,
+                            bool *fallback)
 {
        char link_name[PATH_MAX], *tmp;
        FileEntry fe = FILE_ENTRY__INIT;
@@ -1010,17 +1028,9 @@ again:
        if (ret < 0 && errno == ENOENT) {
                /* Use grand parent, if parent directory does not exist. */
                if (trim_last_parent(link_name) < 0) {
-                       pr_err("trim failed: @%s@\n", link_name);
-                       if (fb && parms->fs_type == OVERLAYFS_SUPER_MAGIC) {
-                               /* In overlayFS, linkat() fails with ENOENT if the removed file is originated from lower layer.
-                                * The cause of failure is that linkat() sees the file has st_nlink=0,
-                                * which is different than st_nlink=1 we got from earlier fstat() on lfd.
-                                * By setting *fb=true, we will fall back to dump_ghost_remap() as it is what should have been
-                                * done to removed files with st_nlink=0.
-                                */
-                               pr_info("Unable to link-remap %s on overlayFS, fall back to dump_ghost_remap\n", path);
-                               *fb = true;
-                       }
+                       check_overlayfs_fallback(params, fallback);
+                       if (*fallback)
+                               pr_err("trim failed: @%s@\n", link_name);
                        return -1;
                }
                goto again;

Reply:
done, please take a look